// office-ui-fabric-react/lib-es2015/components/ScrollablePane/ScrollablePane.styles.js
import { HighContrastSelector, ZIndexes, getGlobalClassNames, } from '../../Styling';
var GlobalClassNames = {
    root: 'ms-ScrollablePane',
};
export var getStyles = function (props) {
    var className = props.className, theme = props.theme;
    var classNames = getGlobalClassNames(GlobalClassNames, theme);
    var AboveAndBelowStyles = {
        position: 'absolute',
        pointerEvents: 'auto',
        width: '100%',
        zIndex: ZIndexes.ScrollablePane
    };
    return ({
        root: [
            classNames.root,
            {
                overflowY: 'auto',
                maxHeight: 'inherit',
                height: 'inherit',
                WebkitOverflowScrolling: 'touch'
            },
            className
        ],
        stickyAbove: [
            {
                top: 0,
                selectors: (_a = {},
                    _a[HighContrastSelector] = {
                        borderBottom: '1px solid WindowText'
                    },
                    _a)
            },
            AboveAndBelowStyles
        ],
        stickyBelow: [
            {
                bottom: 0,
                selectors: (_b = {},
                    _b[HighContrastSelector] = {
                        borderTop: '1px solid WindowText'
                    },
                    _b)
            },
            AboveAndBelowStyles
        ]
    });
    var _a, _b;
};
