// office-ui-fabric-react/lib-es2015/components/Coachmark/Beak/Beak.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { BaseComponent, css, classNamesFunction } from '../../../Utilities';
import { getStyles } from './Beak.styles';
var Beak = /** @class */ (function (_super) {
    tslib_1.__extends(Beak, _super);
    function Beak(props) {
        return _super.call(this, props) || this;
    }
    Beak.prototype.render = function () {
        var _a = this.props, _b = _a.height, height = _b === void 0 ? 18 : _b, _c = _a.width, width = _c === void 0 ? 18 : _c, left = _a.left, top = _a.top;
        var getClassNames = classNamesFunction();
        var classNames = getClassNames(getStyles, {
            left: left,
            top: top,
            height: height + 'px',
            width: width + 'px'
        });
        var pointOne = width / 2 + ',' + 0;
        var pointTwo = width + ',' + height;
        var pointThree = 0 + ',' + height;
        return (React.createElement("div", { className: css('ms-Beak', classNames.root) },
            React.createElement("svg", { height: height, width: width, className: classNames.beak },
                React.createElement("polygon", { points: pointOne + ' ' + pointTwo + ' ' + pointThree }))));
    };
    return Beak;
}(BaseComponent));
export { Beak };
